cocotb/share/lib/embed/embed.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright cocotb contributors
+// Licensed under the Revised BSD License, see LICENSE for details.
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <embed.h>
+#include <cocotb_utils.h>   // xstr, utils_dyn_open, utils_dyn_sym
+#include <gpi.h>            // gpi_event_t
+#include <cstdlib>          // getenv
+#if ! defined(__linux__) && ! defined(__APPLE__)
+#include <windows.h>        // Win32 API for loading the embed impl library
+#include <string>           // string
+#endif
+
+#ifndef EMBED_IMPL_LIB
+#error "Name of embed implementation library required"
+#else
+#define EMBED_IMPL_LIB_STR xstr(EMBED_IMPL_LIB)
+#endif
+
+
+static void (*_embed_init_python)(void);
+static void (*_embed_sim_cleanup)(void);
+static int (*_embed_sim_init)(int argc, char const * const * argv);
+static void (*_embed_sim_event)(gpi_event_t level, const char *msg);
+
+static bool init_failed = false;
+
+
+#define PRINT_ERROR_WINDOWS(log_fmt) {\
+    LPSTR msg_ptr;\
+    if (FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ALLOCATE_BUFFER, NULL, GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_SYS_DEFAULT), reinterpret_cast<LPSTR>(&msg_ptr), 255, NULL)) {\
+        LOG_ERROR(log_fmt ": %s", msg_ptr);\
+        LocalFree(msg_ptr);\
+    } else {\
+        LOG_ERROR(log_fmt);\
+    }}
+
+
+static void * load_embed_impl_lib()
+{
+#if ! defined(__linux__) && ! defined(__APPLE__)
+    HMODULE hm = NULL;
+
+    // get a handle to the module
+    if (!GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, reinterpret_cast<LPCSTR>(&embed_init_python), &hm))
```

Comment:
https://docs.microsoft.com/en-us/windows/win32/intl/conventions-for-function-prototypes suggests we should be using the version without the `A` suffix:

> New Windows applications should use Unicode to avoid the inconsistencies of varied code pages and for ease of localization. They should be written with generic functions, and should define UNICODE to compile the functions into Unicode functions.

Reply:
The `cocotb_utils` library and the error logging macro here uses `FormatMessageA` since the `LOG_*` functions do not handle `wchar_t` strings. If the W-suffix were selected, logs will break. The MSDN documentation states that W-suffix and A-suffix functions should not be used together.

Right now, cocotb is an ANSI, English-only project. Fixing this will require a large internationalization PR and this suggestion would be a part of that.

Reply:
> Right now, cocotb is an ANSI, English-only project. 

Python is very much not ANSI-only, so even if we're "english only", we should try to avoid failing to install on systems with non-ascii paths.

> since the LOG_* functions do not handle wchar_t strings.

Python interprets the arguments to those functions as UTF8, not as ANSI - so the `A` apis may also not work correctly for characters not in the ASCII range.

Reply:
> we should try to avoid failing to install on systems with non-ascii paths

Perhaps this requires some experimentation since I'm fairly certain this is already the case. The issue isn't in the Python, but in the C code and the Python API code. If we get `wchar_t` and pass it to the Python logger we use the `s` format, but we need the `u` specifier to handle `wchar_t`. To support `wchar_t` in the logger we need more work that is out of scope for this PR. That being the case, we should follow Microsoft's guidance and not mix the two formats.

> so the A apis may also not work correctly for characters not in the ASCII range

"ANSI" doesn't mean anything to me, I think on Windows that means Latin-1 8 bit encoding.

Reply:
>  I think on Windows that means Latin-1 8 bit encoding.

In which case, any latin1 characters will result in a `UnicodeDecodeError` from python, as the bytes in the latin1 range are used by UTF8 as a marker for a continuation byte - so the behavior is already broken on ANSI data too.

Using the `W` versions of these functions and `WideCharToMultiByte` to convert the wchar_t to UTF8 would prevent this error. We then just document that the `gpi_log` API takes utf8-encoded strings, which is already the case today.

Reply:
> If we get wchar_t and pass it to the Python logger we use the s format, but we need the u specifier to handle wchar_t.

I don't know what you mean by that. In `str.__mod__`, `%s` and `%u` are synonymous. `PyUnicode_FromString` is how we currently construct the log message, and that uses UTF8 input.

Reply:
> any latin1 characters will result in a UnicodeDecodeError from python, as the bytes in the latin1 range are used by UTF8 as a marker for a continuation byte

Right... so using `A` isn't sufficient either. So it sounds like you want me to make it an action to update to use the `W` suffix unconditionally. And translate all occurrences into UTF-8 before passing to the logger. 

> In str.__mod__, %s and %u are synonymous.

This isn't Python, in C they take two different types: `char*` and `wchar_t*`. Passing `wchar_t` to `s` specifier in Python C API code will most certainly break as it will reinterpret the byte sequence as `char`s. But that doesn't matter now since we will be manually calling out a conversion function.

Reply:
> Right... so using A isn't sufficient either.

Sadly not - otherwise I would probably have accepted your argument.

> Passing wchar_t to s specifier in Python C API code will most certainly break as it will reinterpret the byte sequence as chars.

Where is this `s` specifier in our code that you're referring to?

Reply:
I guess there is no specific use of `s` for logging in a `BuildTuple` or similar that I thought existed (it's been a while since I've looked at that code, I made these changes in June I think). But `gpi_log` and friends only work on UTF-8 and cannot accept `wchar_t*` or UTF-16 directly. If we translate it before passing, then nbd. That's what I had though you were suggesting originally.

Reply:
"translate before passing" sounds like the right approach to me.

Reply:
Thanks to Tim's patches, we don't have to worry about this anymore for this PR. Opened #2273.